icu4c/source/common/uloc.cpp

```diff
@@ -1312,6 +1317,140 @@ _getVariant(const char *localeID,
+    if (variant != nullptr) { variantSink.emplace(variant); }
+
+    ulocimp_getSubtags(
+            localeID,
+            languageSink.has_value() ? &languageSink.value() : nullptr,
+            scriptSink.has_value() ? &scriptSink.value() : nullptr,
+            regionSink.has_value() ? &regionSink.value() : nullptr,
+            variantSink.has_value() ? &variantSink.value() : nullptr,
+            pEnd,
+            status);
+}
+
+U_EXPORT void U_EXPORT2
+ulocimp_getSubtags(
+        const char* localeID,
+        ByteSink* language,
+        ByteSink* script,
+        ByteSink* region,
+        ByteSink* variant,
+        const char** pEnd,
+        UErrorCode& status) {
+    if (U_FAILURE(status)) { return; }
+
+    if (pEnd != nullptr) {
+        *pEnd = localeID;
+    } else if (language == nullptr &&
+               script == nullptr &&
+               region == nullptr &&
+               variant == nullptr) {
+        return;
+    }
+
+    const char* const origLocaleID = localeID;
+    bool hasRegion = false;
+
+    if (localeID == nullptr) {
+        localeID = uloc_getDefault();
+    }
+
+    {
+        CharString tmp = ulocimp_getLanguage(localeID, &localeID, status);
+        if (U_FAILURE(status)) { return; }
+        U_ASSERT(localeID != nullptr);
+        if (language != nullptr) { language->Append(tmp.data(), tmp.length()); }
+    }
+
+    if (pEnd != nullptr) {
+        *pEnd = localeID;
+    } else if (script == nullptr &&
+               region == nullptr &&
+               variant == nullptr) {
+        return;
+    }
+
+    if (_isIDSeparator(*localeID)) {
+        const char* begin = localeID + 1;
+        const char* end = nullptr;
+        CharString tmp = ulocimp_getScript(begin, &end, status);
+        if (U_FAILURE(status)) { return; }
+        U_ASSERT(end != nullptr);
+        if (end != begin) {
+            localeID = end;
+            if (script != nullptr) { script->Append(tmp.data(), tmp.length()); }
+            if (pEnd != nullptr) { *pEnd = localeID; }
+        }
+    }
+
+    if (region == nullptr && variant == nullptr && pEnd == nullptr) { return; }
+
+    if (_isIDSeparator(*localeID)) {
+        const char* begin = localeID + 1;
+        const char* end = nullptr;
+        CharString tmp = ulocimp_getCountry(begin, &end, status);
+        if (U_FAILURE(status)) { return; }
+        U_ASSERT(end != nullptr);
+        if (end != begin) {
+            hasRegion = true;
+            localeID = end;
+            if (region != nullptr) { region->Append(tmp.data(), tmp.length()); }
+            if (pEnd != nullptr) { *pEnd = localeID; }
+        }
+    }
+
+    if (variant == nullptr) { return; }
+
+    CharString tmp;
+    if (_hasBCP47Extension(origLocaleID)) {
+        CharStringByteSink sink(&tmp);
+        ulocimp_forLanguageTag(origLocaleID, -1, sink, nullptr, &status);
+        if (U_SUCCESS(status) && !tmp.isEmpty()) {
+            // This recursive call sets its pEnd to where any variants begin.
```

Comment:
On this item I do want to put my foot down. This is too obscure and surprising even for keeping it only temporarily. Please do add some kind of parameter for stopping after the region code / before the variants.

Reply:
OK, I've now replaced the use of `ulocimp_forLanguageTag()` in `uloc_getVariant()` with a little more straightforward (and much less wasteful) code that instead of converting the entire language tag just checks for the `-t-`, `-u-` or `-x-` marker that indicates the start of a BCP-47 extension.